### PR TITLE
Internship position 262

### PR DESCRIPTION
**Pull request by shkurata**
Ref: #262

### CODE REVIEW

File: api/src/schema/types/internship.ts [hunk 1]
```diff
@@ -128,6 +99,50 @@ builder.queryField('myInternshipInstances', (t) =>
   }),
 );
 
+builder.queryField('availablePositionsForInternshipInstance', (t) =>
```
Comment (moeriki) [inline api/src/schema/types/internship.ts:102]: There's no extra root query needed. The entry point is `myInternshipInstances`. The rest of the data is available through that root query.

**Review by moeriki — CHANGES_REQUESTED** [review]

File: api/src/schema/types/internship.ts (hunk 1 above)
Comment (shkurata) [reply api/src/schema/types/internship.ts:102]: So, we fetch all the instances and filter them on the page?

File: api/src/codegen/schema.graphql [hunk 2]
```diff
@@ -131,11 +132,10 @@ type InternshipInstance implements Node {
 }
 
 type InternshipPosition implements Node {
-  description: String!
   id: ID!
-  mentor: Person!
+  internshipInstance: InternshipInstance!
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/codegen/schema.graphql]: Fixed.

File: api/src/codegen/schema.graphql [hunk 3]
```diff
@@ -303,11 +300,23 @@ type Query {
...
   myInternshipInstances: [InternshipInstance!]!
 }
 
+union QueryAvailablePositionsForInternshipInstanceResult =
+    NotFoundError
+  | QueryAvailablePositionsForInternshipInstanceSuccess
+  | UnauthorizedError
+
+type QueryAvailablePositionsForInternshipInstanceSuccess {
+  data: [InternshipPosition!]!
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/codegen/schema.graphql]: Updated.

File: api/src/domain/internship.ts [hunk 4]
```diff
@@ -35,3 +41,62 @@ export async function getInternshipById(id: string): Promise<InternshipModel> {
...
+    if (internship == null) {
+      throw new AppNotFoundError();
+    }
+    return internship.availablePositions as InternshipPositionModel[];
+  } catch (error) {
+    handleServiceError(error);
+  }
+}
+
+export async function getAvailablePositionsForInstance(
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/domain/internship.ts]: This code is removed.

File: api/src/domain/internship.ts [hunk 5]
```diff
@@ -35,3 +41,62 @@ export async function getInternshipById(id: string): Promise<InternshipModel> {
...
+      },
+    });
+
+    if (internshipInstance == null) {
+      throw new AppNotFoundError();
+    }
+
+    if (internshipInstance.studentId !== context.userId) {
+      throw new AppUnauthorizedError();
+    }
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/domain/internship.ts]: I left this code just to restrict an anonymous access to the details page.

File: api/src/domain/internship.ts [hunk 6]
```diff
@@ -35,3 +41,62 @@ export async function getInternshipById(id: string): Promise<InternshipModel> {
...
+
+    return internshipInstance.internshipPositions as InternshipPositionModel[];
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/domain/internship.ts]: Not really. Removed.

File: api/src/schema/types/internship.ts (hunk 1 above)
Comment (shkurata) [reply api/src/schema/types/internship.ts]: Updated.

File: api/prisma/schema.prisma [hunk 7]
```diff
@@ -122,10 +123,35 @@ model InternshipInstance {
...
 
-  internship   Internship @relation(fields: [internshipId], references: [id])
-  internshipId String
-  student      User?      @relation("InternshipInstanceToStudent", fields: [studentId], references: [id])
-  studentId    String?
+  internship          Internship           @relation(fields: [internshipId], references: [id], onDelete: Cascade)
+  internshipId        String
+  student             User?                @relation("InternshipInstanceToStudent", fields: [studentId], references: [id])
+  studentId           String?
+  internshipPositions InternshipPosition[]
```
[the remark this answers is not in the document]
Comment (moeriki) [reply api/prisma/schema.prisma:130]: But `internshipPositions` is still not needed?

File: api/prisma/schema.prisma [hunk 8]
```diff
@@ -122,10 +123,34 @@ model InternshipInstance {
...
+}
+
+model InternshipPosition {
+  id String @id @default(cuid())
+
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+
+  internships        Internship[]
+  internshipInstance InternshipInstance[]
```
Comment (moeriki) [inline api/prisma/schema.prisma:140]: Not needed yet.

File: api/src/schema/types/organisation.ts [hunk 9]
```diff
@@ -21,7 +21,7 @@ export interface OrganisationModel {
 export const Organisation =
   builder.objectRef<OrganisationModel>('Organisation');
 
-export async function createFakeOrganisation(): Promise<OrganisationModel> {
+export async function createFakeOrganisation(): Promise<OrganisationModelFake> {
```
Comment (moeriki) [inline api/src/schema/types/organisation.ts:24]: This is not used anymore. Remove function and fake model.

**Review by moeriki — CHANGES_REQUESTED** [review]

File: api/src/schema/types/organisation.ts (hunk 9 above)
Comment (shkurata) [reply api/src/schema/types/organisation.ts]: Removed.

File: api/prisma/schema.prisma (hunk 8 above)
Comment (shkurata) [reply api/prisma/schema.prisma]: Removed.

**Review by moeriki — APPROVED** [review]

**Pull request merged**